README.md
Solid Queue Library
===================

About
-----
Solid_queue is a queue with persistent disk storage written in C.
It uses [eblob](https://github.com/reverbrain/eblob) library as backend and provides push/pull interface.
Queue has finite size and if more data come to queue the oldest one will be pushed out from queue.

Usage
-----
### Init
```c
queue_param_t queue_param;
queue_param.eblob_param.blob_size_limit = 1500000000;
queue_param.eblob_param.blob_size = 200000000;
queue_param.eblob_param.records_in_blob = queue_param.eblob_param.blob_size/14000;
queue_param.eblob_param.sync = 5;
queue_param.eblob_param.defrag_timeout = 12;
queue_param.eblob_param.defrag_percentage = 25;
queue_param.eblob_param.blob_flags = EBLOB_TIMED_DATASORT;
queue_param.eblob_param.path = argv[++arg_index];
queue_param.time_to_wait = 5;
queue_param.max_queue_length = 20000;

if(!(queue = queue_open(queue_param)))
{
	exit(1);
}
```

### Pull from queue:
```c
uint64_t len = 0; // size of data from queue - will be set
void *data; // data pulled from queue
queue_pull(queue, &data, &len)
```
### Push to queue:
```c
uint64_t len = 100; // size of data to write
char data[100]; // data to write
bool was_overwrite = false; // if we run out of items free in queue element could be push out
queue_push(queue, data, len, &was_overwrite)
```

### Finalize
```c
queue_close(param->queue);
```

 Create debian package guide
==========================
1. Preparing
------------
To create debian package you need to install some tools. You can do this using command:
```shell
$ sudo apt-get install build-essential dh-make bzr-builddeb
```

After that prepare a tar.gz archive of the source directory. Archive should be called as <name-of-package>-<version>.tar.gz.
For example:
```shell
$ tar -zcvf solid-queue-1.0.1.tar.gz solid_queue
```

Package names must consist only of lower case letters (a-z), digits (0-9), plus (+) and minus (-) signs, and periods (.). They must be at least two characters long and must start with an alphanumeric character.
Then you should make such command: bzr dh_make <name-of-package> <version> <name-of-existing-source-archive>.tar.gz. For example:
```shell
$ bzr dh-make solid-queue 1.0.1 solid-queue-1.0.1.tar.gz
```
It asks type of package. Choose "l" for library package.

2. Editing debian directory
---------------------------
```shell
$ cd solid-queue/debian
```
This directory contains configuration files needed for debian packaging. Most of the files it adds are only needed for specialist packages (such as Emacs modules) so you can start by removing the optional example files:
```shell
$ rm *ex *EX
```

3. Build package
----------------
After editing configuration files use command:
```shell
$ debuild -us -uc
```

References
------
 * EBLOB project https://github.com/reverbrain/eblob
 * Repository to get binary packages of eblob library for your OS: http://repo.reverbrain.com/


debian/changelog
solid-queue (1.3.0.3-1) unstable; urgency=low

  * spec file: depend on eblob-devel >= 0.23.0

 -- Kirill Bushminkin <kbushminkin@gmail.com>  Fri, 31 Jul 2015 10:35:00 +0400
 
solid-queue (1.3.0.2-1) unstable; urgency=low

  * typo in spec file version

 -- Kirill Bushminkin <kbushminkin@gmail.com>  Fri, 31 Jul 2015 10:35:00 +0400
 
solid-queue (1.3.0.1-1) unstable; urgency=low

  * logger: fixed logger initialization
  * depends on eblob version 0.23.0 or higher

 -- Dmitry Rudnev <rudneff.d@gmail.com>  Thu, 30 Jul 2015 10:35:00 +0400

solid-queue (1.3.0.0-1) unstable; urgency=low

  * fixed iterator according to eblob 0.22.23 format

 -- Kirill Bushminkin <kbushminkin@gmail.com>  Wed, 01 Jul 2015 17:09:59 +0400

solid-queue (1.2.0.6-1) unstable; urgency=low

  * fixed version in spec

 -- Kirill Bushminkin <kbushminkin@gmail.com>  Tue, 30 Jun 2015 15:26:00 +0300

solid-queue (1.2.0.5-1) unstable; urgency=low

  * depend on eblob version 0.22.22 or higher

 -- Kirill Bushminkin <kbushminkin@gmail.com>  Tue, 30 Jun 2015 15:26:00 +0300

solid-queue (1.2.0.4-1) unstable; urgency=low

  * depend on eblob version 0.22.22 or higher

 -- Kirill Bushminkin <kbushminkin@gmail.com>  Tue, 30 Jun 2015 15:26:00 +0300

solid-queue (1.2.0.3-1) unstable; urgency=low

  * depend on eblob version 0.22.22 or higher

 -- Kirill Bushminkin <kbushminkin@gmail.com>  Tue, 30 Jun 2015 15:26:00 +0300

solid-queue (1.2.0.2-1) unstable; urgency=low

  * fixed version number in spec
  * fixed eblob version dependency

 -- Kirill Bushminkin <kbushminkin@gmail.com>  Tue, 30 Jun 2015 15:26:00 +0300
 
solid-queue (1.2.0.1-1) unstable; urgency=low

  * init: changed periodic_time to 60 seconds

 -- Dmitry Rudnev <rudneff.d@gmail.com>  Fri, 3 Apr 2015 15:26:00 +0300
 
solid-queue (1.2.0.0-1) unstable; urgency=low

  * init: added a pointer to a custom log handler
  * init: added pointer to pass to log function as first argument

 -- Dmitry Rudnev <rudneff.d@gmail.com>  Tue, 10 Mar 2015 21:20:31 +0300

solid-queue (1.1.0.0-1) unstable; urgency=low

  * init: added error number returning to initial functions
  * test: added some tests
  * deb: added files for debian packaging

 -- Dmitry Rudnev <rudneff.d@gmail.com>  Tue, 16 Dec 2014 17:20:31 +0300

solid-queue (1.0.0.0-1) unstable; urgency=low

  * Initial release

 -- Dmitry Rudnev <rudneff.d@gmail.com>  Tue, 11 Nov 2014 15:07:31 +0300


tests/test_solid_queue.c
/*
 * Copyright 2014+ Dmitry Rudnev <rudneff.d@gmail.com>
 *
 * This file is part of Solid_queue.
 *
 * Solid_queue is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Solid_queue is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Solid_queue.  If not, see <http://www.gnu.org/licenses/>.
 */
#if __STDC_VERSION__ >= 199901L
#define _XOPEN_SOURCE 700
#else
#define _XOPEN_SOURCE 500
#endif /* __STDC_VERSION__ */

#ifdef __GNUC__
#  define UNUSED(x) UNUSED_ ## x __attribute__((__unused__))
#else
#  define UNUSED(x) UNUSED_ ## x
#endif


#define __STDC_FORMAT_MACROS
#include <inttypes.h>
#include <stdlib.h>
#include <stdio.h>
#include <check.h>
#include <unistd.h>
#include <ftw.h>
#include <pthread.h>
#include <errno.h>
#include "solid_queue.h"
 
struct parameters_t
{
	int write_quantity;
	int read_quantity;
	int counter;
	solid_queue_t *queue;
};

pthread_t tid[3];

void* writing(void *p)
{
	struct parameters_t *parameters = (struct parameters_t*)p;
	if(!parameters || !parameters->queue)
	{
		return NULL;
	}
	bool was_overwrite = false;
	for(int i = 0; i < parameters->write_quantity; ++i)
	{
		char ch = 'a';
		queue_push(parameters->queue, &ch, sizeof(char), &was_overwrite);
	}
	return NULL;
}

void* reading(void *p)
{
	struct parameters_t *parameters = (struct parameters_t*)p;
	if(!parameters || !parameters->queue)
	{
		return NULL;
	}
	parameters->counter = 0;
	for(int i = 0; i < parameters->read_quantity; ++i)
	{
		size_t len;
		void *data;
		if(queue_pull(parameters->queue, &data, &len) == 0)
		{
			if(*(char*)data == 'a')
			++parameters->counter;
			free(data);
		}
	}
	return NULL;
}

int unlink_cb(const char *fpath, const struct stat* UNUSED(sb), int UNUSED(typeflag), struct FTW* UNUSED(ftwbuf))
{
	int rv = remove(fpath);

	if (rv)
		perror(fpath);

	return rv;
}

int rmrf(char *path)
{
	return nftw(path, unlink_cb, 64, FTW_DEPTH | FTW_PHYS);
}

int init_test_queue(solid_queue_t **queue, int queue_max_length)
{
	if(*queue) return EINVAL;
	char *temp_name = NULL;
	if(!(temp_name = (char *) malloc(22)))
	{
		return ENOMEM;
	}
	strncpy(temp_name, "/tmp/tmpblobXXXXXX", 21);
	queue_param_t queue_param;
	memset(&queue_param, 0, sizeof(queue_param));

	queue_param.eblob_param.blob_size_limit = 200000000;
	queue_param.eblob_param.blob_size = 20000000;
	queue_param.eblob_param.records_in_blob = queue_param.eblob_param.blob_size/20000;
	queue_param.eblob_param.sync = 5;
	queue_param.eblob_param.defrag_timeout = 12;
	queue_param.eblob_param.defrag_percentage = 25;
	queue_param.eblob_param.blob_flags = EBLOB_TIMED_DATASORT;
	queue_param.max_queue_length = (uint64_t)queue_max_length;
	queue_param.time_to_wait = 10;
	if(!mkdtemp(temp_name))
	{
		return errno;
	}
	queue_param.eblob_param.path = temp_name;
	queue_param.eblob_param.log_level = EBLOB_LOG_ERROR;
	queue_param.eblob_param.log_h = NULL;
	queue_param.eblob_param.priv = NULL;
	int err = queue_open(queue, queue_param);
	return err;
}

START_TEST(test_push_to_queue)
{
	solid_queue_t *solid_queue = NULL;
	int err = 0;
	err = init_test_queue(&solid_queue, 100);
	ck_assert_msg(err == 0, "failed to init queue. Error returned: %i.", err);
	void *data;
	uint64_t len;
	bool was_overwrite = 0;
	err = queue_push(solid_queue, "a", 2, &was_overwrite);
	ck_assert_msg(err == 0, "push failed. Error returned %i.", err);
	queue_pull(solid_queue, &data, &len);
	ck_assert_msg(strcmp(data, "a") == 0, "failure. Wrong returned data: %s.", data);
	ck_assert_msg(len == 2, "failure. Wrong returned length of data.");
	queue_close(solid_queue);
}
END_TEST

START_TEST(test_queue_length)
{
	solid_queue_t *q = NULL;
	int err = 0;
	err = init_test_queue(&q, 4);
	ck_assert_msg(err == 0, "failed to init queue. Error returned: %i.", err);
	void *data;
	uint64_t len;
	bool was_overwrite = 0;
	int i = 0;
	for(i = 0; i < 3; i++)
	{
		err = queue_push(q, "a", 2, &was_overwrite);
		ck_assert_msg(err == 0, "push failed. Error returned %i.", err);
	}
	ck_assert_msg(queue_length(q) == 3, "queue lenght mismatch, after pushed 3 items got length %i.", queue_length(q));
	for(i = 0; i < 2; i++)
	{
		err = queue_pull(q, &data, &len);
		ck_assert_msg(err == 0, "pull failed. Error returned %i.", err);
	}
	ck_assert_msg(queue_length(q) == 1, "queue lenght mismatch, after pulled 2 items got length %i", queue_length(q));
	for(i = 0; i < 6; i++)
	{
		err = queue_push(q, "a", 2, &was_overwrite);
		ck_assert_msg(err == 0, "push failed. Error returned %i.", err);
	}
	ck_assert_msg(queue_length(q) == 4, "queue lenght mismatch, after pushed more then size got length %i", queue_length(q));
	queue_close(q);

}
END_TEST

START_TEST(test_pushes_to_queue)
{
	solid_queue_t *solid_queue = NULL;
	int err = 0;
	err = init_test_queue(&solid_queue, 5);
	ck_assert_msg(err == 0, "failed to init queue. Error returned: %i.", err);
	void *data;
	uint64_t len;
	bool was_overwrite = 0;
	err = queue_push(solid_queue, "a", 2, &was_overwrite);
	ck_assert_msg(err == 0, "push failed. Error returned %i", err);
	err = queue_push(solid_queue, "bb", 3, &was_overwrite);
	ck_assert_msg(err == 0, "push failed. Error returned %i", err);
	err = queue_push(solid_queue, "ccc", 4, &was_overwrite);
	ck_assert_msg(err == 0, "push failed. Error returned %i", err);

	queue_pull(solid_queue, &data, &len);
	ck_assert_msg(strcmp(data, "a") == 0, "failure. Wrong returned data.");
	ck_assert_msg(len == 2, "failure. Wrong returned length of data.");
	free(data);
	data = NULL;

	queue_pull(solid_queue, &data, &len);
	ck_assert_msg(strcmp(data, "bb") == 0, "failure. Wrong returned data.");
	ck_assert_msg(len == 3, "failure. Wrong returned length of data.");
	free(data);
	data = NULL;

	queue_pull(solid_queue, &data, &len);
	ck_assert_msg(strcmp(data, "ccc") == 0, "failure. Wrong returned data.");
	ck_assert_msg(len == 4, "failure. Wrong returned length of data.");
	free(data);
	data = NULL;
}
END_TEST

START_TEST(test_of_thread_safety)
{
	int err = 0;
	struct parameters_t *param = (struct parameters_t*) malloc(sizeof(struct parameters_t));
	ck_assert_msg(param != NULL, "failed to allocate memory.");
	memset(param, 0, sizeof(struct parameters_t));
	err = init_test_queue(&(param->queue), 700);
	ck_assert_msg(err == 0, "failed to init queue. Error returned: %i.", err);
	param->write_quantity = 256;
	param->read_quantity = 512;
	if(pthread_create(&tid[0], NULL, writing, param) != 0 ||
	   pthread_create(&tid[1], NULL, writing, param) != 0 ||
	   pthread_create(&tid[2], NULL, reading, param) != 0)
	{
		ck_abort_msg("failure. Pthread creating.");
	}
	pthread_join(tid[0], NULL);
	pthread_join(tid[1], NULL);
	pthread_join(tid[2], NULL);
	ck_assert_msg(param->counter == param->read_quantity, "failure. Bad read count: %i\n", param->counter);
	queue_close(param->queue);
	free(param);
}
END_TEST

Suite * queue_suite(void)
{
	Suite *s;
	TCase *tc_core;

	s = suite_create("Solid_queue");

	/* Core test case */
	tc_core = tcase_create("Core");

	tcase_add_test(tc_core, test_push_to_queue);
	tcase_add_test(tc_core, test_pushes_to_queue);
	tcase_add_test(tc_core, test_queue_length);
	tcase_add_test(tc_core, test_of_thread_safety);
	suite_add_tcase(s, tc_core);

	return s;
}

int main()
{
	int number_failed;
	Suite *s;
	SRunner *sr;

	s = queue_suite();
	sr = srunner_create(s);

	srunner_run_all(sr, CK_NORMAL);
	number_failed = srunner_ntests_failed(sr);
	srunner_free(sr);
	return (number_failed == 0) ? EXIT_SUCCESS : EXIT_FAILURE;
}


src/solid_queue.c
/*
 * Copyright 2014+ Dmitry Rudnev <rudneff.d@gmail.com>
 * Copyright 2014+ Kirill Bushminkin <kbushminkin@gmail.com>
 *
 * This file is part of library solid_queue.
 *
 * Solid_queue is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Solid_queue is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Solid_queue.  If not, see <http://www.gnu.org/licenses/>.
 */

/**
 * @file solid_queue.c
 * @author Dmitry Rudnev
 * @date 18 Oct 2014
 * @brief File contains implementation of persistent queue interface.
 */
#if __STDC_VERSION__ >= 199901L
#define _XOPEN_SOURCE 600
#else
#define _XOPEN_SOURCE 500
#endif /* __STDC_VERSION__ */

#ifdef __GNUC__
#  define UNUSED(x) UNUSED_ ## x __attribute__((__unused__))
#else
#  define UNUSED(x) UNUSED_ ## x
#endif

#define __STDC_FORMAT_MACROS
#include <inttypes.h>
#include <errno.h>
#include <semaphore.h>
#include <time.h>
#include "solid_queue.h"

#define DEFAULT_PERIODIC_TIMEOUT 60

/**
 * @brief Pointer to iterator handler typedef.
 */
typedef int (*iter_f)(struct eblob_disk_control *dc,
					  struct eblob_ram_control *ctl,
					  int fd,
					  uint64_t data_offset,
					  void *priv,
					  void *thread_priv);

struct _max_min_t
{
	uint64_t max, min;
};

/**
 * @brief Queue structure.
 */
struct _solid_queue_t
{
	struct eblob_backend *eback;
	uint64_t first_key;             /**< Number of first element in queue. */
	uint64_t last_key;              /**< Number of last element in queue. */
	uint64_t length;                /**< Quantity of records in queue. */
	pthread_mutex_t use_queue;      /**< Mutex allowed use queue for only one thread. */
	uint64_t count_remaining;       /**< Count of free cells in queue. */
	sem_t lock_on_empty;            /**< Semaphore prohibits to pull from empty queue. */
	int time_to_wait;               /**< Timer is used for semaphore and mutex timedlock. */
};

uint64_t get_thread_id()
{
	pthread_t ptid = pthread_self();
	return (uint64_t) ptid;
}

/**
 * @brief Iterator handler.
 */
int iterator_h(struct eblob_disk_control *dc,
			   struct eblob_ram_control* UNUSED(ctl),
			   int UNUSED(fd), uint64_t UNUSED(data_offset),
			   void *priv,
			   void* UNUSED(thread_priv))
{
	uint64_t id;
	memcpy(&id, &(dc->key), sizeof(id));
	struct _max_min_t *t= (struct _max_min_t*)priv;
	if(t->max < id && id != 0)
	{
		t->max = id;
	}
	if(t->min > id && id != 0)
	{
		t->min = id;
	}
	return 0;
}

int eblob_log_init (struct eblob_log **el, int level, void *priv, log_handler_t log_h)
{
	if(*el) return EINVAL;
	if(!(*el = (struct eblob_log*) malloc (sizeof(struct eblob_log))))
	{
		return ENOMEM;
	}
	(*el)->log_level = level;
	(*el)->log_private = priv;
	if (!log_h)
	{
		(*el)->log = eblob_log_raw_formatted;
	}
	else
	{
		(*el)->log = log_h;
	}
	return 0;
}


int eblob_config_init(struct eblob_config **econf, const eblob_param_t eblob_param, struct eblob_log *log)
{
	if(*econf) return EINVAL;
	if(!(*econf = (struct eblob_config *) malloc (sizeof(struct eblob_config))))
	{
		return ENOMEM;
	}
	char *fullpath = NULL;
	if(!(fullpath = (char*) malloc (strlen(eblob_param.path) + strlen("/data\0"))))
	{
		free(econf);
		return ENOMEM;
	}
	strcpy(fullpath, eblob_param.path);
	strcat(fullpath, "/data\0");
	(*econf)->blob_flags = eblob_param.blob_flags;
	(*econf)->file = fullpath;
	(*econf)->blob_size = eblob_param.blob_size;
	(*econf)->blob_size_limit = eblob_param.blob_size_limit;
	(*econf)->records_in_blob = eblob_param.records_in_blob;
	(*econf)->log = log;
	(*econf)->sync = eblob_param.sync;
	(*econf)->defrag_percentage = eblob_param.defrag_percentage;
	(*econf)->defrag_timeout = eblob_param.defrag_timeout;
	(*econf)->periodic_timeout = DEFAULT_PERIODIC_TIMEOUT;
	return 0;
}

void eblob_config_free(struct eblob_config *econf)
{
	if(econf)
	{
		free(econf->file);
		free(econf);
	}
}

int iterate_queue(struct _solid_queue_t *q, iter_f iter_func, struct eblob_log *l)
{
	if(!q || !iter_func)
	{
		return -EFAULT;
	}
	q->first_key = 0;
	q->last_key = 0;
	struct eblob_iterate_control iter;
	memset(&iter, 0, sizeof(iter));
	iter.b = q->eback;
	iter.log = l;
	iter.iterator_cb.iterator = iter_func;

	struct _max_min_t max_min;
	memset(&max_min, 0, sizeof(struct _max_min_t));
	max_min.max = 0;
	max_min.min = UINT64_MAX;

	iter.priv = &max_min;
	iter.flags = EBLOB_ITERATE_FLAGS_ALL | EBLOB_ITERATE_FLAGS_READONLY;
	int err = 0;
	err = eblob_iterate(q->eback, &iter);
	if(err != 0)
	{
		return err;
	}
	if(max_min.min != UINT64_MAX)
	{
		q->first_key = max_min.min;
		q->last_key = max_min.max;
		q->length = max_min.max - max_min.min + 1;
	}
	else
	{
		q->first_key = 1;
		q->last_key = 0;
		q->length = 0;
	}
	return 0;
}

int mutex_init(pthread_mutex_t *mutex)
{
	return pthread_mutex_init(mutex, NULL);
}

void mutex_destroy(struct _solid_queue_t *queue)
{
	pthread_mutex_destroy(&(queue->use_queue));
}

int queue_open(struct _solid_queue_t **queue, const queue_param_t queue_param)
{
	if(!queue_param.eblob_param.path)
	{
		return EINVAL;
	}
	struct eblob_log *elog = NULL;
	struct eblob_config *econf = NULL;
	int err = 0;

	if (!(*queue = (struct _solid_queue_t*) malloc (sizeof(struct _solid_queue_t))))
	{
		return ENOMEM;
	}

	if((err = eblob_log_init(&elog, queue_param.eblob_param.log_level, queue_param.eblob_param.priv,
								queue_param.eblob_param.log_h)) != 0 ||
	   (err = eblob_config_init(&econf, queue_param.eblob_param, elog)) != 0 ||
	   !((*queue)->eback = eblob_init(econf)) ||
	   (err = mutex_init(&((*queue)->use_queue))) != 0 ||
	   (err = iterate_queue(*queue, iterator_h, elog)) != 0 ||
	   (err = sem_init(&((*queue)->lock_on_empty), 1, (*queue)->length)) != 0)
	{
		mutex_destroy(*queue);
		if(elog)
			free(elog);
		if(econf)
			eblob_config_free(econf);
		if(*queue)
			free(*queue);
		return (err != 0) ? err : errno;
	}
	(*queue)->count_remaining = ((*queue)->last_key == 0) ? queue_param.max_queue_length :
							 (queue_param.max_queue_length - (*queue)->last_key +
							 (*queue)->first_key - 1);
	(*queue)->time_to_wait = queue_param.time_to_wait;
	eblob_config_free(econf);
	return 0;
}

int push_data(struct _solid_queue_t* queue, void* data, uint64_t len)
{
	++queue->last_key;
	struct eblob_key k;
	memset(&k, 0, sizeof(k));
	memcpy(&k, &(queue->last_key), sizeof(queue->last_key));
	int err = 0;
	if((err = eblob_write(queue->eback, &k, data, 0, (uint64_t)len, 0)) != 0)
	{
		--queue->last_key;
		return err;
	}
	--queue->count_remaining;
	++queue->length;
	return 0;
}

int push_with_displacement(struct _solid_queue_t* queue, void* data, uint64_t len, bool* was_overwrite)
{
	void *d = NULL;
	uint64_t l = 0;
	int err = 0;
	struct eblob_key k;
	memset(&k, 0, sizeof(k));
	memcpy(&k, &(queue->first_key), sizeof(queue->first_key));
	if((err = eblob_read_data(queue->eback, &k, 0, (char**) &d, &l)) != 0)
	{
		return err;
	}
	else
	{
		err = eblob_remove(queue->eback, &k);
		if(err != 0)
		{
			return err;
		}
		++queue->first_key;
		++queue->count_remaining;
		--queue->length;
		if(d) free(d);
	}

	if((err = push_data(queue, data, len)) != 0)
	{
		return err;
	}
	*was_overwrite = 1;
	return 0;
}

int set_timeout(struct timespec* ts, int sec)
{
	if(!ts)
	{
		return -EFAULT;
	}
	if(clock_gettime(CLOCK_REALTIME, ts) == -1)
	{
		return errno;
	}
	ts->tv_sec += sec;

	return 0;
}

int queue_push(struct _solid_queue_t* queue, void* data, uint64_t len, bool *was_overwrite)
{
	struct timespec ts;
	int err = 0;
	int thread_err = 0;
	*was_overwrite = 0;
	if((err = set_timeout(&ts, queue->time_to_wait)) != 0)
	{
		return err;
	}
	if(!queue || !queue->eback)
	{
		return -EFAULT;
	}
	if((thread_err = pthread_mutex_timedlock(&(queue->use_queue), &ts)) != 0)
	{
		return thread_err;
	}
	err = 0;
	if(queue->count_remaining == 0)
	{
		err = push_with_displacement(queue, data, len, was_overwrite);
	}
	else
	{
		err = push_data(queue, data, len);
	}
	if(err == 0 && *was_overwrite == 0 && sem_post(&(queue->lock_on_empty)) == -1)
	{
		return errno;
	}
	if((thread_err = pthread_mutex_unlock(&(queue->use_queue))) != 0)
	{
		return thread_err;
	}
	return err;
}

int queue_pull(struct _solid_queue_t* queue, void **data, uint64_t *len)
{
	struct timespec ts;
	int err = 0;
	int thread_err = 0;
	if((err = set_timeout(&ts, queue->time_to_wait)) != 0)
	{
		return err;
	}
	if(!queue || !data)
	{
		return -EFAULT;
	}
	if(sem_timedwait(&(queue->lock_on_empty), &ts) == -1)
	{
		return errno;
	}
	if((err = pthread_mutex_timedlock(&(queue->use_queue), &ts)) != 0)
	{
		if(sem_post(&(queue->lock_on_empty)) == -1)
		{
			return errno;
		}
		return err;
	}
	*data = NULL;
	*len = 0;
	struct eblob_key k;
	memset(&k, 0, sizeof(k));
	memcpy(&k, &(queue->first_key), sizeof(queue->first_key));
	if((err = eblob_read_data(queue->eback, &k, 0,(char**) data, len)) == 0)
	{
		err = eblob_remove(queue->eback, &k);
		if(err != 0)
		{
			if(sem_post(&(queue->lock_on_empty)) != 0)
			{
				return errno;
			}
			if((thread_err = pthread_mutex_unlock(&(queue->use_queue))) != 0)
			{
				return thread_err;
			}
			return err;
		}
		++queue->first_key;
		++queue->count_remaining;
		--queue->length;
	}
	else
	{
		if(sem_post(&(queue->lock_on_empty)) != 0)
		{
			return errno;
		}
		if((thread_err = pthread_mutex_unlock(&(queue->use_queue))) != 0)
		{
			return thread_err;
		}
		return err;
	}
	if((thread_err = pthread_mutex_unlock(&(queue->use_queue))) != 0)
	{
		return thread_err;
	}
	return 0;
}

void queue_close(struct _solid_queue_t *queue)
{
	if(queue && queue->eback)
	{
		sem_destroy(&(queue->lock_on_empty));
		mutex_destroy(queue);
		eblob_cleanup(queue->eback);
		free(queue);
	}
	return;
}

uint64_t queue_length(struct _solid_queue_t *queue)
{
	return queue->length;
}


example/main.c
/*
 * Copyright 2014+ Dmitry Rudnev <rudneff.d@gmail.com>
 *
 * This file is part of Solid_queue.
 *
 * Solid_queue is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Solid_queue is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Leser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Solid_queue.  If not, see <http://www.gnu.org/licenses/>.
 */
#if __STDC_VERSION__ >= 199901L
#define _XOPEN_SOURCE 600
#else
#define _XOPEN_SOURCE 500
#endif /* __STDC_VERSION__ */

#define __STDC_FORMAT_MACROS
#include <inttypes.h>
#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <assert.h>
#include <stdlib.h>
#include <time.h>
#include "solid_queue.h"

struct perf_stat
{
	volatile uint32_t written_count;
	volatile uint32_t read_count;
	volatile uint32_t write_error_count;
	volatile uint32_t read_error_count;
} perf_stat_instance;

struct parameters_t
{
	int quantity;
	int data_size;
	int data_variance;
	solid_queue_t *queue;
};

volatile uint32_t cnt = 0;
unsigned char is_exiting = 0;
pthread_t tid[5];

void* perf_stat_thread(void *t)
{
	int total = *((int*)t);
	unsigned int last_written = 0;
	unsigned int last_read = 0;

	while(!is_exiting)
	{
		unsigned int write_speed = perf_stat_instance.written_count - last_written;
		unsigned int read_speed = perf_stat_instance.read_count - last_read;

		double percentage = ((double)(perf_stat_instance.written_count +
									  perf_stat_instance.write_error_count)/
									  (double)total) * 100;
		printf("Stat: W: done: %f ", percentage);
		printf("recs: %i wps: %i ", perf_stat_instance.written_count, write_speed);
		printf("err: %i ", perf_stat_instance.write_error_count);
		percentage = ((double)(perf_stat_instance.read_count +
							   perf_stat_instance.read_error_count)/
							   (double)total) * 100;
		printf("R: done: %f reads: %i ", percentage, perf_stat_instance.read_count);
		printf("rps: %i err: %i\n", read_speed, perf_stat_instance.read_error_count);
		last_written = perf_stat_instance.written_count;
		last_read = perf_stat_instance.read_count;
		sleep(1);
	}
	return NULL;
}

void* write_thread(void *p)
{
	struct parameters_t *parameters = (struct parameters_t*)p;
	if(!parameters || !parameters->queue)
	{
		return NULL;
	}
	int max_size_of_data = parameters->data_size + parameters->data_variance;
	char testval[max_size_of_data];
	for(int i = 0; i < max_size_of_data; ++i)
	{
		testval[i] = 'a';
	}
	testval[max_size_of_data-1] = '\0';
	int min_size_of_data = parameters->data_size - parameters->data_variance;
	int difference = parameters->data_variance * 2;
	srand(time(NULL));
	bool was_overwrite = 0;
	for(int i = 0; i < parameters->quantity; ++i)
	{
		if(queue_push(parameters->queue, testval, min_size_of_data + (rand() % difference), &was_overwrite) != 0 )
		{
			__sync_fetch_and_add(&(perf_stat_instance.write_error_count), 1);
			printf("Write thread %" PRIu64 ": going sleep!\n", get_thread_id());
			sleep(3);
		}
		else
		{
			__sync_fetch_and_add(&(perf_stat_instance.written_count), 1);
		}
	}
	return NULL;
}

void* read_thread(void *p)
{
	struct parameters_t *parameters = (struct parameters_t*)p;
	if(!parameters || !parameters->queue)
	{
		return NULL;
	}

	for(int i = 0; i < parameters->quantity; ++i)
	{
		size_t len;
		void *data;
		if(queue_pull(parameters->queue, &data, &len) != 0)
		{
			__sync_fetch_and_add(&(perf_stat_instance.read_error_count), 1);
			printf("Read thread %" PRIu64 " : Error. Going sleep!\n", get_thread_id());
			sleep(3);
		}
		else
		{
			free(data);
			__sync_fetch_and_add(&(perf_stat_instance.read_count), 1);
		}
	}
	return NULL;
}

int main(int argc, char **argv)
{
	if(argc != 5)
	{
		printf("Using: {path_to_queue} {num_of_records} {size_of_rec} {variance_of_rec}\n");
		exit(1);
	}
	struct parameters_t *param = (struct parameters_t*) malloc (sizeof(struct parameters_t));
	if(!param)
	{
		printf("Error while creating parameters.\n");
		exit(1);
	}

	int arg_index = 0;
	queue_param_t queue_param;
	memset(&queue_param, 0, sizeof(queue_param));
	queue_param.eblob_param.blob_size_limit = 1500000000;
	queue_param.eblob_param.blob_size = 200000000;
	queue_param.eblob_param.records_in_blob = queue_param.eblob_param.blob_size/14000;
	queue_param.eblob_param.sync = 5;
	queue_param.eblob_param.defrag_timeout = 12;
	queue_param.eblob_param.defrag_percentage = 25;
	queue_param.eblob_param.blob_flags = EBLOB_TIMED_DATASORT;
	queue_param.eblob_param.path = argv[++arg_index];
	queue_param.time_to_wait = 5;
	queue_param.max_queue_length = 20000;
	param->quantity = atoi(argv[++arg_index]);
	param->data_size = atoi(argv[++arg_index]);
	param->data_variance = atoi(argv[++arg_index]);

	/*Logging init*/
	queue_param.eblob_param.log_level = EBLOB_LOG_ERROR;
	queue_param.eblob_param.log_h = NULL;
	queue_param.eblob_param.priv = NULL;

	if((queue_open(&(param->queue), queue_param)) != 0)
	{
		exit(1);
	}
	if(pthread_create(&tid[0], NULL, perf_stat_thread, &(param->quantity)) != 0 ||
	   pthread_create(&tid[1], NULL, write_thread, param) != 0 ||
	   pthread_create(&tid[2], NULL, read_thread, param) != 0 ||
	   pthread_create(&tid[3], NULL, write_thread, param) != 0 ||
	   pthread_create(&tid[4], NULL, read_thread, param) != 0)
	{
		printf("Can't create thread.\n");
		exit(1);
	}

	pthread_join(tid[1], NULL);
	pthread_join(tid[2], NULL);
	pthread_join(tid[3], NULL);
	pthread_join(tid[4], NULL);
	is_exiting = 1;
	pthread_join(tid[0], NULL);
	queue_close(param->queue);
	free(param);
	printf("All done.\n");

	return 0;
}

